FirebaseInstallations/Source/Library/FIRInstallations.m

```diff
@@ -143,6 +143,40 @@ + (void)validateAppOptions:(FIROptions *)appOptions appName:(NSString *)appName
             kFIRLoggerInstallations, kFIRInstallationsMessageCodeInvalidFirebaseAppOptions,
             [missingFields componentsJoinedByString:@", "]];
   }
+
+  [self validateAPIKey:appOptions.APIKey];
+}
+
++ (void)validateAPIKey:(nullable NSString *)APIKey {
+  NSMutableArray<NSString *> *validationIssues = [[NSMutableArray alloc] init];
+
+  if (APIKey.length != kExpectedAPIKeyLength) {
+    [validationIssues addObject:[NSString stringWithFormat:@"API Key length must be %lu characters",
+                                                           (unsigned long)kExpectedAPIKeyLength]];
+  }
+
+  if (![[APIKey substringToIndex:1] isEqualToString:@"A"]) {
+    [validationIssues addObject:@"API Key must start with `A`"];
+  }
+
+  NSCharacterSet *characters = [NSCharacterSet characterSetWithCharactersInString:APIKey];
+  if (![[NSCharacterSet alphanumericCharacterSet] isSupersetOfSet:characters]) {
+    [validationIssues addObject:@"API Key must contain only alphanumeric characters"];
+  }
+
+  if (validationIssues.count > 0) {
+    NSLog(@"---%@---", APIKey);
+    [NSException
+         raise:kFirebaseInstallationsErrorDomain
+        format:
+            @"%@[%@] Could not configure Firebase Installations due to invalid FirebaseApp "
+            @"options. `FirebaseOptions.APIKey` doesn't match the expected format: %@. If you use "
```

Comment:
I wonder if this sentence is helpful ("`FirebaseOptions.APIKey` doesn't match the expected format: %@").

I recommend we say 
* the API key is invalid
* please update the `GoogleServices-Info.plist` file.

Afaik, that is the only recommended way to configure Firebase apps. \
Yes, developers are free to programmatically initialize Firebase options - but at their own risk.

Reply:
I would prefer to provide all failure information we know. It may help us with debugging an issue in the future.

Reply:
My gut-feeling tells me that sharing syntactical implementation details with developers can create confusion.
Sharing the semantics of the error however can make it clear what is going wrong.

(exaggerated) Example:
syntactical error message: "_The format of your input String is too long._"
semantical error message with action item: "_You provided an invalid API key. Please download the latest `GoogleServices-Info.plist` file from the Firebase Console to solve the issue._"

Since we don't agree, I would recommend to reach out to a tech writer.
The tech writer for FIS is usually egilmore@. wdyt?

Reply:
@egilmorez Do you have an opinion on it? Please let us know if you would like to change anything - I'll create a separate PR for it.